### DISCUSSION TITLE
How to conduct out-of-sample prediction

### DISCUSSION THREAD

**Discussion by MichaelHopwood** [pyg-team/pytorch_geometric#5201] [Q&A]
Is there an established way to utilize GraphSage's inductive properties to make an inference on a new graph (e.g. the same graph in training but with more nodes added)?

The examples I see in the `examples/` directory utilize an index to cut up a single graph for train & test sets.

As a note - I'd want to apply this for either a node classification problem or an unsupervised node embedding problem.

**Comment by rusty1s** [comment]
Yes, you should be able to re-use the same node but on a modified graph. For this, you can simply create a new loader that now operates on your new data. Let me know if this helps you get started!

**Reply by MichaelHopwood** [reply]
@rusty1s Are we allowed to conduct out-of-sample prediction in this way?

```py
#model = trained node classification GCN model with inductive NeighborLoader setting on Cora dataset
#        used `data.subgraph(data.train_mask)` in NeighborLoader

from torch_geometric.data import Data
x=dataset[0].x
# adding a new node which shares same features as node 43
inference_x = torch.cat((x,x[42,:]),0)

A = dataset[0].edge_index
# adding graph connections for this new node
inference_adj = torch.cat((A, 
               torch.LongTensor([[2708,24],[2708,245],[2708,622]]).T
               ), 1)

inference_Data = Data(x=inference_x, edge_index=inference_adj)

# in-sample prediction
y_hat = model(dataset[0])
print(y_hat)

# out-of-sample prediction
y_hat = model(inference_Data)
print(y_hat)
```

**Reply by rusty1s** [reply]
Yes, this looks good. Please note that you also need to add reverse edges as well:
```python
[24, 2708], [245, 2708], [622, 2708]
```

**Reply by MichaelHopwood** [reply]
@rusty1s Thanks for your response. Is there anything inhibiting inference on a completely new data?

For instance, I trained a model on my data and then did an inference on Cora. Other than mandating that the node features maintains the same feature dimension, it seemed to work fine.

```py
# model = trained on my dataset, using inductive & NeighborLoader

# Start inference
dataset_cora = get_planetoid_dataset('cora')
dataset = [Data(x=dataset_cora[0].x[:, :41].clone(),
                y=dataset_cora[0].y,
                edge_index=dataset_cora[0].edge_index)]
preds = model(dataset[0])
# works without error
```
Does doing this raise any red flags?
Is this how transfer learning is conducted in GNNs?

**Reply by rusty1s** [reply]
I am not sure if this makes a lot of sense. The domain and features in your data need to match with the ones the model was trained on. Otherwise, this just corresponds to an untrained model.

**Reply by MichaelHopwood** [reply]
Yeah sorry I just meant this as a proof of concept. Certainly it doesn't make sense here but if I have (for instance) a graph which updates every few days and shares the same feature set, then I suppose I would use this method to perform inductive inference on the new graph & nodes.

**Reply by rusty1s** [reply]
Yes, that is correct.
